src/gc.c

```diff
@@ -191,11 +221,35 @@ size_t gc_NumericIndex(RedisModuleCtx *ctx, GarbageCollectorCtx *gc) {
     goto end;
   }
 
-  FieldSpec *randomField = numericFields[rand() % array_len(numericFields)];
+  if (array_len(numericFields) != array_len(gc->numericGCCtx)) {
+    // add all numeric fields to our gc
+    assert(array_len(numericFields) >
+           array_len(gc->numericGCCtx));  // it is not possible to remove fields
+    gc_FreeNumericGcCtxArray(gc);
+    for (int i = 0; i < array_len(numericFields); ++i) {
+      NumericRangeTree *rt = OpenNumericIndex(sctx, numericFields[i]->name, &idxKey);
+      // if we could not open the numeric field we probably have a
+      // corruption in our data, better to know it now.
+      assert(rt);
+      gc->numericGCCtx = array_append(gc->numericGCCtx, gc_NewNumericGcCtx(rt));
+      if (idxKey) RedisModule_CloseKey(idxKey);
+    }
+  }
+
   array_free(numericFields);
-  NumericRangeTree *rt = OpenNumericIndex(sctx, randomField->name, &idxKey);
 
-  NumericRangeNode *nextNode = NextGcNode(rt);
+  // choose random numeric gc ctx
+  int randomIndex = rand() % array_len(gc->numericGCCtx);
+  NumericFieldGCCtx *numericGcCtx = gc->numericGCCtx[randomIndex];
+  if (numericGcCtx->revisionId != numericGcCtx->rt->revisionId) {
+    // revision changed, recreating our numeric gc ctx
```

Comment:
I think it's probably safer here to not hold on to the pointer of `rt`, but rather just open it anew, using the field name. Do we have a guarantee that the pointer is still valid?

Reply:
It will not change since we do not remove fields and I thought it will be more efficient this way...

Reply:
We are also holds the tree nodes in the numeric iterator and its impossible to reopen those, so I do not see any reason why not saving the tree pointer ...

Reply:
The tree nodes are dependent on each other. I'm worried about the scenario where we drop the index and the gc is still running. But the fix is a simpler one: Simply get the spec outside gcRandomTerm, and then return if the spec is not found. There is also the potential issue of delete and immediately restore. Basically, we cannot assume that the spec is not deleted, and that if the index still exists, it is the exact same index as was initially created.

Reply:
Actually this line is doing what you said and I got it on gc_NumericIndex function:
RedisSearchCtx *sctx = NewSearchCtx(ctx, (RedisModuleString *)gc->keyName);

But you are correct it could be that someone deleted it and recreated it immediately and then I need to spot it and recreate the iterators. The only way I see that we can spot it is adding a an id to each spec and then check if the spec id changed.

Is there any other way you think we can solve it?

Reply:
That's what I was thinking about exactly. The id doesn't need to be persistent; it's just updated whenever the spec changes, so that everything is rescanned. Like a revision ID.

Should we save that for later?

Reply:
Yes I think it will be better to save it for later, it not that big task but we need to make sure its done correctly.